backend/selectgen.ml

```diff
@@ -1047,13 +1112,18 @@ method emit_expr (env:environment) exp =
               (* Ccatch registers must not contain out of heap pointers *)
               Array.iter (fun reg -> assert(reg.typ <> Addr)) src;
               self#insert_moves env src tmp_regs ;
-              self#insert_moves env tmp_regs (Array.concat dest_args) ;
-              begin match env_close_regions env dest_regions with
-              | None -> ()
-              | Some regs -> self#insert env (Iop Iendregion) regs [||]
+              self#insert_moves env tmp_regs (Array.concat handler.regs) ;
+              if handler.regions != env.regions then begin
+                let cl, rest =
+                  cut_region_stack
+                    ~len:(List.length handler.regions)
+                    env.regions
+                in
+                assert (rest == handler.regions);
```

Comment:
It had better! If we exit out of a tail back into a region after doing local allocs in the tail, we'll interleave allocations from two different regions.

Reply:
I think that isn't entirely implied by having this condition fail. I think that means your assumption is that a handler is executed in the same region as when you enter its body. That does indeed seem a reasonable assumption, especially for the weird staticcatch constructs with fall-through that these IRs have instead of proper continuations. I'd be less sure that flambda2 will obey that rule, but it doesn't use these block-structured tail constructs anyway for precisely this kind of reason.

Reply:
Actually, it's more than that -- I think the entire approach to dealing with static-handlers in this file relies on that assumption, so clearly this assertion is fine.